File: smartedu-backend/src/main/java/com/smartedu/controller/StudentController.java
package com.smartedu.controller;

import com.smartedu.model.Student;
import com.smartedu.service.StudentService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@CrossOrigin(origins = "*", maxAge = 3600)
@RestController
@RequestMapping("/api/students")
public class StudentController {

    @Autowired
    private StudentService studentService;

    @PostMapping
    @PreAuthorize("hasRole('ADMIN') or hasRole('TEACHER')")
    public ResponseEntity<Student> createStudent(@RequestBody Student student) {
        return ResponseEntity.ok(studentService.saveStudent(student));
    }

    @GetMapping
    @PreAuthorize("hasRole('ADMIN') or hasRole('TEACHER')")
    public ResponseEntity<List<Student>> getAllStudents() {
        return ResponseEntity.ok(studentService.getAllStudents());
    }

    @GetMapping("/{id}")
    @PreAuthorize("hasRole('ADMIN') or hasRole('TEACHER')")
    public ResponseEntity<Student> getStudentById(@PathVariable Long id) {
        return ResponseEntity.ok(studentService.getStudentById(id));
    }

    @GetMapping("/user/{userId}")
    @PreAuthorize("hasRole('ADMIN') or hasRole('TEACHER') or hasRole('STUDENT')")
    public ResponseEntity<Student> getStudentByUserId(@PathVariable Long userId) {
        return ResponseEntity.ok(studentService.getStudentByUserId(userId));
    }

    @GetMapping("/me")
    @PreAuthorize("hasRole('STUDENT')")
    public ResponseEntity<Student> getMyStudentProfile() {
        com.smartedu.security.services.UserDetailsImpl userDetails = 
            (com.smartedu.security.services.UserDetailsImpl) org.springframework.security.core.context.SecurityContextHolder.getContext().getAuthentication().getPrincipal();
        return ResponseEntity.ok(studentService.getStudentByUserId(userDetails.getId()));
    }

    @PutMapping("/{id}")
    @PreAuthorize("hasRole('ADMIN') or hasRole('TEACHER')")
    public ResponseEntity<Student> updateStudent(@PathVariable Long id, @RequestBody Student studentDetails) {
        return ResponseEntity.ok(studentService.updateStudent(id, studentDetails));
    }

    @DeleteMapping("/{id}")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<?> deleteStudent(@PathVariable Long id) {
        studentService.deleteStudent(id);
        return ResponseEntity.ok().build();
    }
}

File: smartedu-backend/controllers/studentController.js
const studentService = require('../services/studentService');

async function createStudent(req, res, next) {
    try {
        const student = await studentService.saveStudent(req.body);
        return res.json(student);
    } catch (error) {
        next(error);
    }
}

async function getAllStudents(req, res, next) {
    try {
        const students = await studentService.getAllStudents();
        return res.json(students);
    } catch (error) {
        next(error);
    }
}

async function getStudentById(req, res, next) {
    try {
        const { id } = req.params;
        const student = await studentService.getStudentById(id);
        return res.json(student);
    } catch (error) {
        next(error);
    }
}

async function getStudentByUserId(req, res, next) {
    try {
        const { userId } = req.params;
        const student = await studentService.getStudentByUserId(userId);
        return res.json(student);
    } catch (error) {
        next(error);
    }
}

async function getMyStudentProfile(req, res, next) {
    try {
        const userId = req.user.id;
        const student = await studentService.getStudentByUserId(userId);
        return res.json(student);
    } catch (error) {
        next(error);
    }
}

async function updateStudent(req, res, next) {
    try {
        const { id } = req.params;
        const student = await studentService.updateStudent(id, req.body);
        return res.json(student);
    } catch (error) {
        next(error);
    }
}

async function deleteStudent(req, res, next) {
    try {
        const { id } = req.params;
        await studentService.deleteStudent(id);
        return res.json({ success: true, message: 'Student deleted successfully' });
    } catch (error) {
        next(error);
    }
}

module.exports = {
    createStudent,
    getAllStudents,
    getStudentById,
    getStudentByUserId,
    getMyStudentProfile,
    updateStudent,
    deleteStudent
};
